pkg/coscheduling/core/core.go

```diff
@@ -238,7 +271,102 @@ func (pgMgr *PodGroupManager) calculateAssignedPods(podGroupName, namespace stri
+		if len(insufficientResources) > 0 {
+			continue
+		}
+		return &leftResources, nil
+	}
+	return &leftResources, insufficientResources
+}
+
 // GetNamespacedName returns the namespaced name
 func GetNamespacedName(obj metav1.Object) string {
 	return fmt.Sprintf("%v/%v", obj.GetNamespace(), obj.GetName())
 }
+
+func resourceSatisfied(leftResource, req *framework.Resource) []noderesources.InsufficientResource {
+	insufficientResources := make([]noderesources.InsufficientResource, 0, 4)
+	if leftResource.Memory < req.Memory {
+		insufficientResources = append(insufficientResources, noderesources.InsufficientResource{
+			ResourceName: corev1.ResourcePods,
+			Reason:       "Insufficient Memory",
+		})
+	}
+	if leftResource.MilliCPU < req.MilliCPU {
+		insufficientResources = append(insufficientResources, noderesources.InsufficientResource{
+			ResourceName: corev1.ResourceCPU,
+			Reason:       "Insufficient CPU",
+		})
+	}
+	if leftResource.EphemeralStorage < req.EphemeralStorage {
+		insufficientResources = append(insufficientResources, noderesources.InsufficientResource{
+			ResourceName: corev1.ResourceEphemeralStorage,
+			Reason:       "Insufficient ephemeral storage",
+		})
+	}
+	// AllowedPodNumber seems not correct, so we need check again
+	if leftResource.AllowedPodNumber < req.AllowedPodNumber {
+		insufficientResources = append(insufficientResources, noderesources.InsufficientResource{
+			ResourceName: corev1.ResourcePods,
+			Reason:       "Too many pods",
+		})
+	}
+	for rName, rQuant := range req.ScalarResources {
+		if leftResource.ScalarResources[rName] < rQuant+leftResource.ScalarResources[rName] {
+			insufficientResources = append(insufficientResources, noderesources.InsufficientResource{
+				ResourceName: rName,
+				Reason:       fmt.Sprintf("Insufficient %v", rName),
+			})
+		}
+	}
+	return insufficientResources
+}
+
+func getNodeResource(info *framework.NodeInfo) *framework.Resource {
+	leftResource := framework.Resource{
+		ScalarResources: make(map[corev1.ResourceName]int64),
+	}
+	allocatable := info.Allocatable
+	requested := info.Requested
+
+	leftResource.AllowedPodNumber = allocatable.AllowedPodNumber - len(info.Pods)
+	leftResource.MilliCPU = allocatable.MilliCPU - requested.MilliCPU
+	leftResource.Memory = allocatable.Memory - requested.Memory
+	leftResource.EphemeralStorage = allocatable.EphemeralStorage - requested.EphemeralStorage
+
+	for k, allocatableEx := range allocatable.ScalarResources {
+		requestEx, ok := requested.ScalarResources[k]
+		if !ok {
+			leftResource.ScalarResources[k] = allocatableEx
+		} else {
+			leftResource.ScalarResources[k] = allocatableEx - requestEx
+		}
+	}
+	klog.V(4).Infof("Node %v left resource %+v", info.Node().Name, leftResource)
+	return &leftResource
+}
+
+func formatErrors(notFits []noderesources.InsufficientResource) error {
+	err := fmt.Errorf("Cluster resourc not enough")
+	for _, notFit := range notFits {
+		err = errors.Wrap(err, notFit.Reason)
```

Comment:
Is it mandatory to introduce this `github.com/pkg/errors` dependency? If not, I'd like to keep it simple and only use built-in Go library.